### PR TITLE
Add initial implementation of extracting pb blocks from ddisasm output

### PR DESCRIPTION
**Pull request by junghee**

### CODE REVIEW

File: disassemblers/ddisasm/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# DDISASM
+
+[ddsiasm](https://git.grammatech.com/rewriting/ddisasm)
```
Comment (aeflores) [inline disassemblers/ddisasm/README.md]:
```suggestion
[ddisasm](https://github.com/GrammaTech/ddisasm)
```

File: disassemblers/ddisasm/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+## Requirement
+
+- [gtirb](https://github.com/grammatech/gtirb)
```
Comment (aeflores) [inline disassemblers/ddisasm/README.md]:
```suggestion
- [gtirb](https://github.com/GrammaTech/gtirb)
```

File: disassemblers/ddisasm/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+- [gtirb-pprinter](https://git.grammatech.com/rewriting/gtirb-pprinter)
```
Comment (aeflores) [inline disassemblers/ddisasm/README.md]:
```suggestion
```
I don't think this code depends on the pprinter, does it?

File: disassemblers/ddisasm/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+- [ddisasm](https://git.grammatech.com/rewriting/ddisasm)
```
Comment (aeflores) [inline disassemblers/ddisasm/README.md]:
```suggestion
- [ddisasm](https://github.com/GrammaTech/ddisasm)
```

File: disassemblers/ddisasm/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+
+```
+python3 -m pip install requirements.txt
+```
+
+## Usage
+
+```
+PYTHONPATH=../../protobuf_def python3 ddisasmBBnRef.py -i [input-binary or gtirb-ir-file] -o [output-path]
```
Comment (aeflores) [inline disassemblers/ddisasm/README.md]:
```suggestion
PYTHONPATH=../../protobuf_def python3 ddisasmBBorRef.py -i [input-binary or gtirb-ir-file] -o [output-path]
```

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    """
+    try:
+        p = subprocess.run(
+            ["ddisasm", binary, "--ir", output, "-j", "1"], check=True
+        )
+    except subprocess.CalledProcessError as e:
+        raise DdisasmError(f"ddisasm failed on {binary}: {e}")
+
+
+def addBB(block, pbFunc):
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]:
```suggestion
def addBB(block:gtirb.CodeBlock, pbFunc):
```
what is the type of pbFunc?

File: disassemblers/ddisasm/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+gtirb
+gtirb-pprinter
```
Comment (aeflores) [inline disassemblers/ddisasm/requirements.txt]: I don't think this code requires the pprinter
```suggestion

```

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+            pbFunc = module.fuc.add()
+            pbFunc.va = entry_block.address
+
+            for block in m.aux_data["functionBlocks"].data[function]:
+                addBB(block, pbFunc)
+
+        logging.info("Writing output to %s", output)
+        f = open(output, "wb")
+        f.write(module.SerializeToString())
+        f.close()
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]:
```suggestion
        with open(output, "wb") as f:
            f.write(module.SerializeToString())
```

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            entries = m.aux_data.get("functionEntries").data[function]
+            logging.debug("function: %s", func_name)
+
+            # Pick the lowest address if there are multiple entries
+            entry_block = min(entries)
+
+            pbFunc = module.fuc.add()
+            pbFunc.va = entry_block.address
+
+            for block in m.aux_data["functionBlocks"].data[function]:
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py:104]: what happens with basic blocks that don't belong to any function? I am not sure if ddisasm can generate those.
Is there a way to encode basic blocks that don't belong to functions or does the sok representation assume every basic block is inside a function?

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+    for m in ir.modules:
+        logging.info("Module %s", m.name)
+
+        module = blocks_pb2.module()
+        for function, func_name in m.aux_data["functionNames"].data.items():
+            entries = m.aux_data.get("functionEntries").data[function]
+            logging.debug("function: %s", func_name)
+
+            # Pick the lowest address if there are multiple entries
+            entry_block = min(entries)
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py:99]: Does this sort by address by default?

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+    pbBB.va = block.address
+    pbBB.parent = pbFunc.va
+
+    logging.debug(
+        "The basic block address is 0x{0:x}, size is {1}".format(
+            pbBB.va, pbBB.size
+        )
+    )
+
+    decoder = GtirbInstructionDecoder(block.module.isa)
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]: it might make sense to pass this object into the function so you don't have to create one for every block.

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+
+    # get successors
+    for out_edge in block.outgoing_edges:
+        target_block = out_edge.target
+        if not isinstance(target_block, gtirb.ProxyBlock):
+            child = pbBB.child.add()
+            child.va = target_block.address
+            logging.debug("Successor of 0x{0:x}: 0x{1:x}".format(
+                block.address, target_block.address)
+            )
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py:80]: oh, this is orthogonal to our current effort, but can we also get stats on how good our CFG is? that would be interesting to see.

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+                    if symbol._payload is None:
+                        continue
+
+                    if isinstance(symbol._payload, gtirb.Block):
+                        tblock = symbol._payload
+                        if isinstance(tblock, gtirb.ProxyBlock):
+                            # TODO: ??
+                            continue
+
+                        to_addr = tblock.byte_interval.address + tblock.offset
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]:
```suggestion
                        to_addr = tblock.address
```

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+
+                    logging.debug(
+                        "%d: From 0x%x -> 0x%x"
+                        % (kind, ref.ref_va, ref.target_va)
+                    )
+
+    logging.info("Collect Refs done! ready to write output...")
+    f = open(output, "wb")
+    f.write(refInf.SerializeToString())
+    f.close()
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]:
```suggestion
    with open(output, "wb") as f:
        f.write(refInf.SerializeToString())
```

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 15]
```diff
@@ -0,0 +1,305 @@
+import logging
+import optparse
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]: `optparse` is deprecated since python 3.2, why use it? I think we should be using `argparse`

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+class ExtractMode(Enum):
+    BB = "bb"
+    REF = "ref"
+
+
+if __name__ == "__main__":
+    parser = optparse.OptionParser()
+    parser.add_option(
+        "-m", "--mode",
+        metavar="string",
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]: I don't think you need to define a metavar here, if you don't the documentation will be `--mode=MODE` which is fine.

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+    BB = "bb"
+    REF = "ref"
+
+
+if __name__ == "__main__":
+    parser = optparse.OptionParser()
+    parser.add_option(
+        "-m", "--mode",
+        metavar="string", 
+        help="Select extract-mode: `bb` or `ref`",
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py:292]: if you use `argparse` you can ensure only legal options are used
```suggestion
        choices=[ExtractMode.BB.value,ExtractMode.REF.value],
        help="Select extract-mode: `bb` or `ref`",
```

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+        default=ExtractMode.BB.value,
+    )
+    parser.add_option(
+        "-i", "--input",
+        action="store",
+        type="string",
+        help="input binary or gtirb-IR file",
+        default=None
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]: assuming you switch to argparse I think you want
```suggestion
        type=Path,
        help="input binary or gtirb-IR file",
        required=True,
```
same for the output option

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+        action="store",
+        type="string",
+        help="output of the protobuf file",
+        default=None
+    )
+    (options, args) = parser.parse_args()
+
+    if options.input is None:
+        print("Please provide an input binary or gtirb-IR file")
+        exit(-1)
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]: If you use `required=True` above, you don't need to check here.

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+        ir_path = Path(pb_output).with_suffix(".gtirb")
+        disassemble(Path(options.input), ir_path)
+        logging.info("Done with ddisasm.")
+        end = time.monotonic()
+        logging.info(f"time-for-ddisasm: {options.input}: {end - start:.2f}")
+    elif file_type == FileType.IR:
+        ir_path = options.input
+    else:
+        logging.warning(f"Unknown file type: %s", options.input)
+        exit(-1)
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]: why -1? I would usually expect `exit(1)` for something like this. Same for other -1s

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+        pb_output = Path(options.output)
+    else:
+        pb_output = Path("/tmp") / f"{Path(options.input).name}.ddisasm.pb"
+
+    file_type = get_magic_file_type(options.input)
+
+    if file_type == FileType.ELF or file_type == FileType.PE:
+        logging.info("Run ddisasm on %s", options.input)
+        start = time.monotonic()
+        ir_path = Path(pb_output).with_suffix(".gtirb")
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]:
```suggestion
        ir_path = pb_output.with_suffix(".gtirb")
```
pb_output is already a Path object

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 17 above)
Comment (aeflores) [reply disassemblers/ddisasm/ddisasmBBorRef.py:292]: you might even be able to do
```
type = ExtractMode
```
and `argparse` will take care of casting this for you, so you don't have to check later if the string can be transformed to an ExtractMode object (I am not 100% sure of this).

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+                from_addr = bi.address + offset
+
+                for symbol in sym_expr.symbols:
+                    if symbol._payload is None:
+                        continue
+
+                    if isinstance(symbol._payload, gtirb.Block):
+                        tblock = symbol._payload
+                        if isinstance(tblock, gtirb.ProxyBlock):
+                            # TODO: ??
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py]:
```suggestion
                            # This is an external reference which is not captured here.
```

File: disassemblers/ddisasm/ddisasmBBorRef.py [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+        module.isa == gtirb.Module.ISA.IA32
+        or module.isa == gtirb.Module.ISA.ARM
+    ):
+        return 4
+    elif (
+        module.isa == gtirb.Module.ISA.X64
+        or module.isa == gtirb.Module.ISA.ARM64
+    ):
+        return 8
+    else:
```
Comment (aeflores) [inline disassemblers/ddisasm/ddisasmBBorRef.py:139]: I thought we had mips32 too

**Review by aeflores — CHANGES_REQUESTED** [review]

File: disassemblers/ddisasm/README.md (hunk 1 above)
Comment (junghee) [reply disassemblers/ddisasm/README.md]: Fixed.

File: disassemblers/ddisasm/README.md (hunk 2 above)
Comment (junghee) [reply disassemblers/ddisasm/README.md]: Fixed.

File: disassemblers/ddisasm/README.md (hunk 3 above)
Comment (junghee) [reply disassemblers/ddisasm/README.md]: You are right. Removed.

File: disassemblers/ddisasm/README.md (hunk 4 above)
Comment (junghee) [reply disassemblers/ddisasm/README.md]: Fixed.

File: disassemblers/ddisasm/README.md (hunk 5 above)
Comment (junghee) [reply disassemblers/ddisasm/README.md]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 6 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Added blocks_pb2.Function

File: disassemblers/ddisasm/requirements.txt (hunk 7 above)
Comment (junghee) [reply disassemblers/ddisasm/requirements.txt]: Removed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 8 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 9 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py:104]: It looks like dummy pbFunc can be created. I will add it after all the other comments are done addressed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 10 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py:99]: It does not sort by address, and instead, it iterates the list once to determine the smallest element. Why do you ask? am I missing something..?

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 11 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 12 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py:80]: I will revisit this after the first round of review change is done.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 13 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 14 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 15 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 16 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Removed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 17 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py:292]: It looks like it gives an error when we have both due to type conversion error.

I think having choices is better than having type=ExtractMode in that the choices case produces better error message showing the choices.

Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 18 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Hmm,, I thought I would have the output as optional, but I don't really insist. So, fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 19 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Removed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 20 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 21 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 22 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py]: Fixed.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 23 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py:139]: Added.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 9 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py:104]: OK, I just pushed the addition.

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 12 above)
Comment (junghee) [reply disassemblers/ddisasm/ddisasmBBorRef.py:80]: I just realized that this code adding child of each block seems to be enough to compare cfg results. I will try to see if compareCFG.py works with it as it is.

File: disassemblers/ddisasm/README.md [hunk 24]
```diff
@@ -0,0 +1,22 @@
+# DDISASM
+
+[ddsiasm](https://github.com/GrammaTech/ddisasm)
```
Comment (aeflores) [inline disassemblers/ddisasm/README.md:3]:
```suggestion
[ddisasm](https://github.com/GrammaTech/ddisasm)
```

File: disassemblers/ddisasm/ddisasmBBorRef.py (hunk 10 above)
Comment (aeflores) [reply disassemblers/ddisasm/ddisasmBBorRef.py:99]: Sorry, what I meant is that `entries` are code blocks if I remember correctly and I don't know offhand if the default ordering for code blocks is based on addresses. E.g. If it sorts based on UUID then this could be problematic.

Should this be `min(entries, key= lambda x: x.address)` ?

**Review by aeflores — CHANGES_REQUESTED** [review]

**Comment by junghee** [comment]
Fixed in the GT repo
